### PR TITLE
chore: change date handling

### PR DESCRIPTION
**Pull request by BenjaminMoe**
Right now all of the tests are passing except for one, which is the date as an object case which has been giving me trouble each step of the way. 

```
 FAIL  src/__tests__/date.test.ts (35.481s)
  Test 1. Confirm behavior of issuanceDate
    ✓ 1. case-0 should match the verifiable credential (389ms)
    ✓ 1. case-1 should match the verifiable credential (285ms)
    ✓ 1. case-2 should match the verifiable credential (283ms)
    ✓ 1. case-3 should match the verifiable credential (300ms)
    ✓ 1. case-4 should match the verifiable credential (308ms)
    ✓ 1. case-5 should match the verifiable credential (283ms)
    ✓ 1. case-6 should match the verifiable credential (280ms)
    ✓ 1. case-7 should match the verifiable credential (292ms)
    ✓ 1. case-8 should match the verifiable credential (262ms)
    ✓ 1. case-9 should match the verifiable credential (282ms)
    ✓ 1. case-10 should match the verifiable credential (290ms)
    ✓ 1. case-11 should match the verifiable credential (276ms)
    ✓ 1. case-12 should match the verifiable credential (270ms)
    ✓ 1. case-13 should match the verifiable credential (265ms)
    ✓ 1. case-14 should match the verifiable credential (262ms)
    ✓ 1. case-15 should match the verifiable credential (274ms)
    ✓ 1. case-16 should match the verifiable credential (255ms)
    ✓ 1. case-17 should match the verifiable credential (286ms)
    ✓ 1. case-18 should match the verifiable credential (274ms)
    ✓ 1. case-19 should match the verifiable credential (271ms)
    ✓ 1. case-20 should match the verifiable credential (266ms)
    ✓ 1. case-21 should match the verifiable credential (278ms)
    ✓ 1. case-22 should match the verifiable credential (275ms)
    ✓ 1. case-23 should match the verifiable credential (288ms)
    ✓ 1. case-24 should match the verifiable credential (265ms)
    ✓ 1. case-25 should match the verifiable credential (258ms)
    ✓ 1. case-26 should match the verifiable credential (269ms)
    ✓ 1. case-27 should match the verifiable credential (272ms)
    ✓ 1. case-28 should match the verifiable credential (287ms)
    ✕ 1. case-29 should match the verifiable credential (70ms)
  Test 2. Confirm behavior of suite date constructor
    ✓ 2. case-0 removed should match the verifiable credential (299ms)
    ✓ 2. case-1 undefined should match the verifiable credential (298ms)
    ✓ 2. case-2 null should match the verifiable credential (295ms)
    ✓ 2. case-3 0 should match the verifiable credential (300ms)
    ✓ 2. case-4  should match the verifiable credential (293ms)
    ✓ 2. case-5 foobar should match the verifiable credential (1ms)
    ✓ 2. case-6 1635774995208 should match the verifiable credential (288ms)
    ✓ 2. case-7 -1635774995208 should match the verifiable credential (260ms)
    ✓ 2. case-8 1 should match the verifiable credential (292ms)
    ✓ 2. case-9 -1 should match the verifiable credential (283ms)
    ✓ 2. case-10 123 should match the verifiable credential (377ms)
    ✓ 2. case-11 -123 should match the verifiable credential (272ms)
    ✓ 2. case-12 12345 should match the verifiable credential (247ms)
    ✓ 2. case-13 -12345 should match the verifiable credential (279ms)
    ✓ 2. case-14 1991-08-25T21:33:56+09:00 should match the verifiable credential (277ms)
    ✓ 2. case-15 Sunday, August 25th 1991, 9:33:56 pm should match the verifiable credential (1ms)
    ✓ 2. case-16 25th Sunday August 1991 should match the verifiable credential (1ms)
    ✓ 2. case-17 Sunday August 25, 1991 should match the verifiable credential (264ms)
    ✓ 2. case-18 25 Aug 1991 should match the verifiable credential (275ms)
    ✓ 2. case-19 1991-08-25 should match the verifiable credential (262ms)
    ✓ 2. case-20 Sun, 25 Aug 1991 21:33:56  should match the verifiable credential (258ms)
    ✓ 2. case-21 08 25 1991 should match the verifiable credential (261ms)
    ✓ 2. case-22 Aug 25, 1991 should match the verifiable credential (256ms)
    ✓ 2. case-23 1991-08-25T21:33:56 should match the verifiable credential (275ms)
    ✓ 2. case-24 1991-08-25T21:33:56:789+09:00 should match the verifiable credential (1ms)
    ✓ 2. case-25 1991-08-25T21:33:56Z should match the verifiable credential (261ms)
    ✓ 2. case-26 1991-08-25T21:33+09:00 should match the verifiable credential (254ms)
    ✓ 2. case-27 1991-08-25T12:33:56.789Z should match the verifiable credential (266ms)
    ✓ 2. case-28 1991,7,25,21,33,56,789 should match the verifiable credential (1ms)
    ✓ 2. case-29 [object Object] should match the verifiable credential
  Test 3. Confirm behavior of suite date set directly
    ✓ 3. case-0 removed should match the verifiable credential (262ms)
    ✓ 3. case-1 undefined should match the verifiable credential (276ms)
    ✓ 3. case-2 null should match the verifiable credential (253ms)
    ✓ 3. case-3 0 should match the verifiable credential (269ms)
    ✓ 3. case-4  should match the verifiable credential (251ms)
    ✓ 3. case-5 foobar should match the verifiable credential (278ms)
    ✓ 3. case-6 1635774995208 should match the verifiable credential (272ms)
    ✓ 3. case-7 -1635774995208 should match the verifiable credential (265ms)
    ✓ 3. case-8 1 should match the verifiable credential (271ms)
    ✓ 3. case-9 -1 should match the verifiable credential (273ms)
    ✓ 3. case-10 123 should match the verifiable credential (275ms)
    ✓ 3. case-11 -123 should match the verifiable credential (266ms)
    ✓ 3. case-12 12345 should match the verifiable credential (248ms)
    ✓ 3. case-13 -12345 should match the verifiable credential (292ms)
    ✓ 3. case-14 1991-08-25T21:33:56+09:00 should match the verifiable credential (277ms)
    ✓ 3. case-15 Sunday, August 25th 1991, 9:33:56 pm should match the verifiable credential (281ms)
    ✓ 3. case-16 25th Sunday August 1991 should match the verifiable credential (277ms)
    ✓ 3. case-17 Sunday August 25, 1991 should match the verifiable credential (284ms)
    ✓ 3. case-18 25 Aug 1991 should match the verifiable credential (276ms)
    ✓ 3. case-19 1991-08-25 should match the verifiable credential (285ms)
    ✓ 3. case-20 Sun, 25 Aug 1991 21:33:56  should match the verifiable credential (272ms)
    ✓ 3. case-21 08 25 1991 should match the verifiable credential (288ms)
    ✓ 3. case-22 Aug 25, 1991 should match the verifiable credential (340ms)
    ✓ 3. case-23 1991-08-25T21:33:56 should match the verifiable credential (295ms)
    ✓ 3. case-24 1991-08-25T21:33:56:789+09:00 should match the verifiable credential (298ms)
    ✓ 3. case-25 1991-08-25T21:33:56Z should match the verifiable credential (283ms)
    ✓ 3. case-26 1991-08-25T21:33+09:00 should match the verifiable credential (297ms)
    ✓ 3. case-27 1991-08-25T12:33:56.789Z should match the verifiable credential (271ms)
    ✓ 3. case-28 1991,7,25,21,33,56,789 should match the verifiable credential (25ms)
    ✓ 3. case-29 [object Object] should match the verifiable credential (3ms)
  Test 4. Confirm behavior of issuanceDate
    ✓ 4. case-0 should match the verifiable credential (306ms)
    ✓ 4. case-1 should match the verifiable credential (294ms)
    ✓ 4. case-2 should match the verifiable credential (307ms)
    ✓ 4. case-3 should match the verifiable credential (303ms)
    ✓ 4. case-4 should match the verifiable credential (275ms)
    ✓ 4. case-5 should match the verifiable credential (1ms)
    ✓ 4. case-6 should match the verifiable credential (312ms)
    ✓ 4. case-7 should match the verifiable credential (362ms)
    ✓ 4. case-8 should match the verifiable credential (356ms)
    ✓ 4. case-9 should match the verifiable credential (305ms)
    ✓ 4. case-10 should match the verifiable credential (289ms)
    ✓ 4. case-11 should match the verifiable credential (347ms)
    ✓ 4. case-12 should match the verifiable credential (325ms)
    ✓ 4. case-13 should match the verifiable credential (417ms)
    ✓ 4. case-14 should match the verifiable credential (317ms)
    ✓ 4. case-15 should match the verifiable credential (2ms)
    ✓ 4. case-16 should match the verifiable credential (1ms)
    ✓ 4. case-17 should match the verifiable credential (344ms)
    ✓ 4. case-18 should match the verifiable credential (307ms)
    ✓ 4. case-19 should match the verifiable credential (323ms)
    ✓ 4. case-20 should match the verifiable credential (275ms)
    ✓ 4. case-21 should match the verifiable credential (291ms)
    ✓ 4. case-22 should match the verifiable credential (366ms)
    ✓ 4. case-23 should match the verifiable credential (497ms)
    ✓ 4. case-24 should match the verifiable credential (1ms)
    ✓ 4. case-25 should match the verifiable credential (498ms)
    ✓ 4. case-26 should match the verifiable credential (511ms)
    ✓ 4. case-27 should match the verifiable credential (463ms)
    ✓ 4. case-28 should match the verifiable credential (1ms)
    ✓ 4. case-29 should match the verifiable credential (1ms)
```

The issue here is that Digital Bazaar throws an error on sign while Transmute allows the object to pass through. The code that does this is pretty hard to trace through, and kind of seems magical in terms of where the error is actually occurring. What kind of compounds this, is that the first test is failing, but the others are passing, so we need to know why. And the Array tests are passing, so we also need to know why.

### CODE REVIEW

**Comment by BenjaminMoe** [comment]
Okay, so for the first point, why is the first test failing, but the following tests are passing. And the reason seems to be the structure of the tests. In the first test we're setting the `issuanceDate` parameter to an object, while not applying any logic to the suite. 

In the following tests, we're passing the date into the suite, and or signature. And we're getting an error in the same place where the date is handled, and then the same 'invalid date' error is being thrown in the same place.

**Comment by BenjaminMoe** [comment]
As for the second point about the array working. In the case `issuanceDate` the array seems to be allowed to pass through for both examples, and then throws the same error on signing examples. So we can say that it's really a problem that comes down to the individual case where the `issuanceDate` is throwing an error in Digital Bazaar's suite. Trying to trace down exactly what the problem is could take a while, so I think it would be a good idea to ask Orie about this one.

**Comment by BenjaminMoe** [comment]
After checking with Orie about what the expected behavior should be, the answer was to use `checkCredential` in the `vc.js` package to check the JSON-LD so we can throw an error similar to that of Digital Bazaar. This function isn't directly available from the package, so first commit will be to expose this functionality.

**Comment by BenjaminMoe** [comment]
Now that `checkCredential` is available from `vc.js` the next step is going to be to import this function into the `ed25519-signature-2018` package so that we can make sure a credential is valid JSON-LD when we sign it. 

I have a feeling that this could be one of those things where it fixes my date test, but then breaks a lot of other stuff. So I want to be careful in terms of unintended consequences. First I'll put `checkCredential` in my date test to see how it affects the results of the test.

Since calling `checkCredential` in the test wouldn't provide an accurate depiction of what the library is actually doing, from there i can put the function into the library, see if the date test works, and then see how the rest of the CI tests still work. And then we can see how we want to proceed from there is depending on how much of an impact is made. 

So next commit will be to install @transmute/vc.js into ed25519, and then add it into the tests.

**Comment by BenjaminMoe** [comment]
Alternative solution would be to skip the object test on the `issuanceDate` tests so that the tests are passing, and then kick the JSON-LD discrepancy into another issue.

**Comment by BenjaminMoe** [comment]
Aaaaaand the CI tests are failing. More reason to not to further address the JSON-LD pull request in the same PR. Changing the behavior of the ED25519-2018 suite probably means that alof of fixtures  are probably going to be different, which means we need to address all of the tests that broke to push this PR through. We'll address that and then create another ticket for the JSON-LD error. 

Next set of commits will be going through the tests and making sure they pass locally.

**Comment by BenjaminMoe** [comment]
I'm going to mark this as ready for review. What's changed is that we handle dates the same way Digital Bazaare does in the ED25519-2018 suite. Overall the consequences of this change shouldn't be too over reaching, as it mainly forces XML date strings for the `created` attribute of proofs, which should be the default behavior already. The main change is that the constructor throws an error when a bad date is encountered, which means we might have to put try/catch statements around the constructor.

File: packages/ed25519-signature-2018/src/Ed25519Signature2018.ts [hunk 1]
```diff
@@ -28,7 +28,12 @@ export class Ed25519Signature2018 {
   public verificationMethod?: string;
   constructor(options: IEd25519Signature2018Options = {}) {
     this.signer = options.signer;
-    this.date = options.date;
+    if (options.date) {
+      this.date = new Date(options.date);
```
Comment (OR13) [inline packages/ed25519-signature-2018/src/Ed25519Signature2018.ts:32]: this converts the date from what is supplied in the option.... if we are doing this for compatibility, we should warn... anything that changes an input provided by the caller should be a warning, or an error.

File: packages/ed25519-signature-2018/src/Ed25519Signature2018.ts [hunk 2]
```diff
@@ -157,15 +162,21 @@ export class Ed25519Signature2018 {
...
-      date = new Date(date).toISOString();
-      date = date.substr(0, date.length - 5) + "Z";
+    if (date && typeof date !== "string") {
+      if (date === undefined || date === null) {
+        date = new Date();
+      } else if (typeof date === "number" || typeof date === "string") {
+        date = new Date(date);
+      }
+      const str = date.toISOString();
+      date = str.substr(0, str.length - 5) + "Z";
```
Comment (OR13) [inline packages/ed25519-signature-2018/src/Ed25519Signature2018.ts:172]: if this line ever happens there should be a warning... again, mutating method arguments should result in a warning or an error.

File: packages/ed25519-signature-2018/src/Ed25519Signature2018.ts [hunk 3]
```diff
@@ -157,15 +162,21 @@ export class Ed25519Signature2018 {
...
     }
 
     // add API overrides
-    if (date !== undefined) {
+    if (date) {
```
Comment (OR13) [inline packages/ed25519-signature-2018/src/Ed25519Signature2018.ts:176]: dislike this type check, what if data is 0, 1, "string" etc...

File: packages/ed25519-signature-2018/src/__tests__/date.test.ts [hunk 4]
```diff
@@ -111,8 +111,8 @@ const signCredential = async (
 const verifyProof = async (document: any, proof: any) => {
   const keyPair = await Ed25519VerificationKey2018.from(rawKeyJson);
   const suite = new Ed25519Signature2018({
-    key: keyPair,
-    date: proof.created
+    key: keyPair
+    //date: proof.created
```
Comment (OR13) [inline packages/ed25519-signature-2018/src/__tests__/date.test.ts:115]: remove comment

**Review by OR13 — CHANGES_REQUESTED** [review]
We should probably chat about the "warn" vs "error" part of this.

**Review by OR13 — APPROVED** [review]

**Pull request merged**